source/array_list_with_map.c

```diff
@@ -0,0 +1,127 @@
+    AWS_PRECONDITION(list_with_map);
+    AWS_PRECONDITION(element);
+    size_t current_length = aws_array_list_length(&list_with_map->list);
+    if (current_length == 0) {
+        /* Nothing to remove */
+        return AWS_OP_SUCCESS;
+    }
+    /* find the index of element first */
+
+    struct aws_hash_element *find = NULL;
+    /* find and remove the elment from table */
+    if (aws_hash_table_find(&list_with_map->map, element, &find)) {
+        return AWS_OP_ERR;
+    }
+    if (!find) {
+        /* It's removed already */
+        return AWS_OP_SUCCESS;
+    }
+
+    size_t index_to_remove = (size_t)find->value;
+    if (aws_hash_table_remove(&list_with_map->map, element, NULL, NULL)) {
+        return AWS_OP_ERR;
+    }
+    /* Nothing else can fail after here. */
+
+    if (index_to_remove != current_length - 1) {
+        /* It's not the last element, we need to swap it with the end of the list and remove the last element */
+        void *last_element = aws_mem_acquire(list_with_map->list.alloc, list_with_map->list.item_size);
+        AWS_FATAL_ASSERT(aws_array_list_back(&list_with_map->list, last_element) == AWS_OP_SUCCESS);
+        /* Update the last element index in the table */
+        struct aws_hash_element *element_to_update = NULL;
+        AWS_FATAL_ASSERT(aws_hash_table_find(&list_with_map->map, last_element, &element_to_update) == AWS_OP_SUCCESS);
+        AWS_FATAL_ASSERT(element_to_update != NULL);
+        element_to_update->value = (void *)index_to_remove;
+        /* Swap the last element with the element to remove in the list */
+        aws_array_list_swap(&list_with_map->list, index_to_remove, current_length - 1);
+        aws_mem_release(list_with_map->list.alloc, last_element);
+    }
+    /* Remove the current last element from the list */
+    AWS_FATAL_ASSERT(aws_array_list_pop_back(&list_with_map->list) == AWS_OP_SUCCESS);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_array_list_with_map_get_random(struct aws_array_list_with_map *list_with_map, void *out) {
+    AWS_PRECONDITION(list_with_map);
+    size_t length = aws_array_list_length(&list_with_map->list);
+    if (length == 0) {
+        return aws_raise_error(AWS_ERROR_LIST_EMPTY);
+    }
+
+    /* use the best of two algorithm to select the connection with the lowest load. */
+    uint64_t random_64_bit_num = 0;
+    aws_device_random_u64(&random_64_bit_num);
+    size_t index = random_64_bit_num % length;
+    AWS_FATAL_ASSERT(aws_array_list_get_at(&list_with_map->list, out, index) == AWS_OP_SUCCESS);
+    return AWS_OP_SUCCESS;
+}
+
+size_t aws_array_list_with_map_length(struct aws_array_list_with_map *list_with_map) {
+    return aws_array_list_length(&list_with_map->list);
+}
+
+bool aws_array_list_with_map_exist(struct aws_array_list_with_map *list_with_map, const void *element) {
+    struct aws_hash_element *find = NULL;
+    AWS_FATAL_ASSERT(aws_hash_table_find(&list_with_map->map, element, &find) == AWS_OP_SUCCESS);
```

Comment:
why is this a fatal assert?

Reply:
it will only fail when the hash table is not valid any more. https://github.com/awslabs/aws-c-common/blob/main/source/hash_table.c#L412. So, I think a fatal assert will be better here